File: site/showcase/android-notes/app/src/main/java/com/example/notes/NotesViewModel.java
package com.example.notes;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

/**
 * Survives configuration changes — maps UI intents to repository calls.
 */
public class NotesViewModel extends ViewModel {
  private final NotesRepository repository;

  public NotesViewModel(NotesRepository repository) {
    this.repository = repository;
  }

  /** Default factory path in the showcase — real apps use a ViewModelFactory. */
  public NotesViewModel() {
    this(new NotesRepository(AppDatabase.getInstance().noteDao()));
  }

  public LiveData<java.util.List<Note>> getNotes() {
    return repository.observeNotes();
  }

  public void addNote(String body) {
    if (body == null || body.isBlank()) return;
    repository.insert(new Note(body.trim()));
  }

  public void refresh() {
    repository.refreshFromNetwork();
  }
}

File: site/showcase/swift-notes/Sources/Notes/NotesViewModel.swift
import Foundation
import Combine

/// Maps UI intents to the repository — ObservableObject for SwiftUI.
@MainActor
final class NotesViewModel: ObservableObject {
  @Published private(set) var notes: [Note] = []

  private let repository: NotesRepository

  init(repository: NotesRepository = NotesRepository()) {
    self.repository = repository
  }

  func load() {
    notes = repository.fetchAll()
  }

  func addNote(_ body: String) {
    let trimmed = body.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else { return }
    repository.insert(Note(body: trimmed))
    load()
  }

  func refresh() {
    repository.refreshFromNetwork()
    load()
  }
}

struct Note: Identifiable {
  let id = UUID()
  let body: String
}
